### ISSUE TITLE
Failure type narrowing unioned `TypeIs`

### ISSUE THREAD

**Issue by tigerhawkvok** [astral-sh/ty#4394]
### Summary

This is the most trivial form of set-theoretic typing trimmed out of one of my codebases. This did _not_ report on `v0.0.56`, and does on `v0.0.74`.

```py
from typing import Any, Never, Union
import pandas as pd
import pyspark.pandas as ps
from typing_extensions import TypeIs

PandasTypeFrame = Union[ps.DataFrame, pd.DataFrame]
AnyOtherType = Never

class _FrameIdentities():
    """
    A collection of static methods to provide
    identity transformations and checks on DataFrames
    of various provenances.
    """

    @staticmethod
    def isCorePandasFrame(frame:Any) -> TypeIs[pd.DataFrame]:
        """
        A syntactic sugar type guard for a core pandas frame
        """
        return isinstance(frame, pd.DataFrame)

    @staticmethod
    def isPandasOnSparkFrame(frame:Any) -> TypeIs[ps.DataFrame]:
        """
        A syntactic sugar type guard for a pandas-on-spark frame
        """
        return isinstance(frame, ps.DataFrame)

    @classmethod
    def isPandasTypeFrame(cls, frame:Any) -> TypeIs[PandasTypeFrame]:
        """
        A type guard for either flavor of pandas frame
        """
        return cls.isCorePandasFrame(frame) or cls.isPandasOnSparkFrame(frame)

    @classmethod
    def pdFrame(cls, frame:Any) -> pd.DataFrame:
        # Snipped irrelevant frame-casting magic
        return pd.DataFrame([{"foo": 1, "bar": 2}])

def returnsPandas(_df:Union[AnyOtherType, PandasTypeFrame]) -> pd.DataFrame:
    if _FrameIdentities.isPandasTypeFrame(_df):
        if _FrameIdentities.isPandasOnSparkFrame(_df):
            # Coerce
            _df = _FrameIdentities.pdFrame(_df)
        # element `pyspark.pandas.frame.DataFrame[Unknown] & ~pyspark.pandas.frame.DataFrame[Unknown]` of union `(pyspark.pandas.frame.DataFrame[Unknown] & ~pyspark.pandas.frame.DataFrame[Unknown]) | pandas.core.frame.DataFrame` is not assignable to `pandas.core.frame.DataFrame` ty(invalid-return-type) // 0.0.74
        return _df
    raise NotImplementedError("Other irrelevant code here")
```

The first `isPandasTypeFrame` guarantees a trivial union of two classes, and the assignment guarded by `.isPandasOnSparkFrame()` guarantees the removal of one of the two classes. The report even makes it clear: `pyspark.pandas.frame.DataFrame[Unknown] & ~pyspark.pandas.frame.DataFrame[Unknown]` is `Never`

`Never|Something` is always `Something`.

If the argument is being made that unknown and unknown don't anti-intersect (a bad argument, IMO, and presumes the generic `TypeIs` deserves "unknown" instead of "Any" -- non-specialization is "everything" not "some poorly defined specific thing that may not be the same") -- specializing the `.isPandasOnSparkFrame()` to `TypeIs[ps.DataFrame[Any]]` _still_ reports an error -- and surely `ps.DataFrame[Any]` is a superset of `ps.DataFrame[Unknown]` and _absolutely_ should reduce to `Never`

### Version

0.0.74

**Comment by carljm** [comment]
Thanks for the report! Here is a simplified reproducer that demonstrates the same behavior, for easier discussion:

```py
from typing import TypeIs

def is_list(value: object) -> TypeIs[list]:
    return isinstance(value, list)

def f(value: list | int) -> int:
    if is_list(value):
        value = 0
    reveal_type(value)  # on current ty, reveals `(list[Unknown] & ~list[Unknown]) | int`
    return value  # fails on current ty
```

We agree this example should narrow successfully, and we’ll use this issue to track a refinement of `TypeIs` narrowing behavior.

Under the current theoretical foundations specified for Python typing, `list[Any] & ~list[Any]` (or `DataFrame[Any] & ~DataFrame[Any]`) is not necessarily `Never`: `Any` and `Unknown` both represent unknown static types, and their separate occurrences can represent different specializations. That is, `list[Any]` does not represent the type "all possible lists", just as `Any` does not represent the type "all possible objects". The type "all possible objects" is spelled `object`. `Any` is "an unknown static type" and `list[Any]` is "some list type, whose specialization is not known".

The type "all possible lists, of any specialization" (in other words, the infinite union `type[T1] | type[T2] | ...` for all possible `Tn`) is called the "top materialization" of `list[Any]`. In narrowing to an invariant generic type, it's important to be able to express this type, and there is currently no way to spell it in the standard Python type system.

ty does provide the necessary spelling:

```py
from typing import TypeIs
from ty_extensions import Top

def is_list(value: object) -> TypeIs[Top[list]]:
    return isinstance(value, list)
```

But this is not portable since it is not yet standardized in the Python type system, and it is hard to use because `ty_extensions` has no runtime shim and so must be protected with `if TYPE_CHECKING` in runtime code. In the future we hope to make `Top` a standardized part of the type system.

In the meantime, we plan to align `TypeIs` narrowing of gradual generic targets with our current behavior for `isinstance`:

- By default, use the top materialization to eliminate alternatives, while preserving compatible known type arguments or the declared gradual specialization in the resulting type. Thus `TypeIs[list[Any]]` can exclude all lists in the negative branch while still producing `list[Any]` when narrowing from `object`. This behavior isn't fully theoretically consistent, but it matches user expectations well.
- With `analysis.strict-generic-narrowing = true`, retain the top materialization in the resulting type, as we used to do before astral-sh/ruff#26864.

This should accept your example without requiring ty-specific annotations. Until then, replacing the inner guard call with `isinstance(_df, ps.DataFrame)` is also a workaround.

The ty behavior here changed in astral-sh/ruff#26864. Previously, we always used the top materialization of a gradual `TypeIs` target. We removed that conversion to preserve explicitly requested gradual types, but this has undesirable effects on the narrowing behavior.
